### PR TITLE
Kv Version Diff View

### PR DESCRIPTION
**Pull request by Monkeychip**
I attempted to improve the logic on this page but ran into familiar issues I had last time I implemented this component. I added some notes to the code. 

Changes from old view:
- green check icon does not indicated selected version, instead it mirrors the version dropdown and version history views and only shows next to the current version.
- If a version is selected it has blue text given by the class name `is-active` on a link. Very similar to what we do on the client counts calendar widget.
- It shows an empty state if the current secret is destroyed or deleted. We default to showing the current version as the base version off of which to compare. 
- It selects the version prior to the current version. If that is deleted or destroyed it selects the next one which is not either of those things. If all secrets are deleted or destroyed, it defaults back to one prior. 

image
image
image

### CODE REVIEW

File: ui/app/routes/vault/cluster/secrets/backend/secret-edit.js [hunk 1]
```diff
@@ -320,7 +320,7 @@ export default Route.extend(UnloadModelRoute, {
       if (!model.hasDirtyAttributes || model.isDeleted) {
         return true;
       }
-      // TODO: below is KV v2 logic, remove with engine work
+      // TODO: below is KV v2 logic, remove with engine work. kv engine cleanup
```
Comment (Monkeychip) [inline ui/app/routes/vault/cluster/secrets/backend/secret-edit.js:323]: Added additional text here for global search on this key phrase `kv engine cleanup` as noted in the JIRA ticket.

File: ui/app/styles/core/json-diff-patch.scss
Comment (Monkeychip) [inline ui/app/styles/core/json-diff-patch.scss:1]: For now, kind of duplicate CSS, but well remove the component css and replace with this core file.

File: ui/lib/kv/addon/components/page/secret/metadata/version-diff.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    const versionData = this.args.metadata.sortedVersions.find(
+      (version) =>
+        version.destroyed === false && version.deletion_time === '' && version.version != this.leftSideVersion
+    );
+    // if all versions have been deleted or destroyed return one less than the current version;
+    return versionData ? versionData.version : this.leftSideVersion - 1;
+  }
+
+  async fetchSecretData(version) {
+    // ARG TODO ask if maybe we want to override the adapter peekRecord?
```
Comment (Monkeychip) [inline ui/lib/kv/addon/components/page/secret/metadata/version-diff.js:47]: As per my comment in the code, originally Chelsea and I had talked about making a method in the adapter that did what I'm doing here. Checking the store and if nothing exists then hitting up queryRecord. We could do this by overriding peekRecord in the data adapter. 

Thoughts?

File: ui/lib/kv/addon/components/page/secret/metadata/version-diff.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    const rightSideVersionData = await this.fetchSecretData(this.rightSideVersion);
+    const delta = diffpatcher.diff(rightSideVersionData, leftSideVersionData);
+
+    this.statesMatch = !delta;
+    this.visualDiff = delta
+      ? jsondiffpatch.formatters.html.format(delta, rightSideVersionData)
+      : JSON.stringify(leftSideVersionData, undefined, 2);
+  }
+
+  @action selectVersion(version, actions, side) {
```
Comment (Monkeychip) [inline ui/lib/kv/addon/components/page/secret/metadata/version-diff.js:70]: I really wanted to just mut these tracked properties in the template, but had massive issues with promises and the order of things in the getters. This still bugs me.

File: ui/lib/kv/addon/components/page/secret/details.hbs
Comment (Monkeychip) [inline ui/lib/kv/addon/components/page/secret/details.hbs:1]: you'll see this added to other components, but we only show the diff view if there is more than one version.

Comment (hellobontempo) [inline ui/lib/kv/addon/components/page/secret/details.hbs:1]: Kind of an aside - seeing these duplicate navigation tabs makes me think that the page header should be rendered one, higher up - then we could just use `{{outlet}}` for these child detail views

File: ui/lib/kv/addon/components/page/secret/metadata/version-diff.hbs [hunk 4]
```diff
@@ -1,11 +1,90 @@
...
+                  {{/if}}
+                </button>
+              </li>
+            {{/each}}
+          </ul>
+        </nav>
+      </D.Content>
+    </BasicDropdown>
+    {{! right side version selector }}
+    <BasicDropdown @class="popup-menu" @horizontalPosition="auto-right" @verticalPosition="below" as |D|>
```
Comment (hellobontempo) [inline ui/lib/kv/addon/components/page/secret/metadata/version-diff.hbs:45]: Wondering if there's a way to pass in an `on click` function and make `KvVersionDropdown reusable`

File: ui/lib/kv/addon/components/page/secret/metadata/version-diff.hbs [hunk 5]
```diff
@@ -1,11 +1,90 @@
...
+              <li>
+                <button
+                  type="button"
+                  class="link"
+                  {{on "click" (fn this.selectVersion versionData.version D.actions "left")}}
+                  disabled={{(or versionData.destroyed versionData.deletion_time)}}
+                >
+                  Version
+                  {{versionData.version}}
+                  {{#if (eq versionData.version this.leftSideVersion)}}
```
Comment (hellobontempo) [inline ui/lib/kv/addon/components/page/secret/metadata/version-diff.hbs:30]: I'm a little confused - are the icons supposed to give us information about the version state? or the diff state?

File: ui/lib/kv/addon/components/page/secret/metadata/version-diff.js [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+ * @param {string} path - Backend from the kv/data model.
+ * @param {array} metadata - The kv/metadata model. It is version agnostic.
+ * @param {array} breadcrumbs - Array to generate breadcrumbs, passed to the page header component.
+ */
+
+export default class KvVersionDiffComponent extends Component {
+  @service store;
+
+  @tracked leftSideVersion = this.args.metadata.currentVersion;
+  @tracked rightSideVersion = this.defaultRightSideVersion;
```
Comment (hellobontempo) [inline ui/lib/kv/addon/components/page/secret/metadata/version-diff.js:27]: I think tracked properties typically should be initially set in the constructor?

File: ui/lib/kv/addon/components/page/secret/metadata/version-diff.js [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+  async createVisualDiff() {
+    /* eslint-disable no-undef */
+    const diffpatcher = jsondiffpatch.create({});
+    const leftSideVersionData = await this.fetchSecretData(this.leftSideVersion);
+    const rightSideVersionData = await this.fetchSecretData(this.rightSideVersion);
+    const delta = diffpatcher.diff(rightSideVersionData, leftSideVersionData);
+
+    this.statesMatch = !delta;
+    this.visualDiff = delta
```
Comment (hellobontempo) [inline ui/lib/kv/addon/components/page/secret/metadata/version-diff.js:64]: could this be a getter that updates based on the left and right sides changing?

File: ui/lib/kv/addon/components/page/secret/metadata/version-diff.hbs (hunk 5 above)
Comment (Monkeychip) [reply ui/lib/kv/addon/components/page/secret/metadata/version-diff.hbs:30]: Version state for destroyed and deleted, but if you've selected the version is has the green icon (circle not filled in).
image

File: ui/lib/kv/addon/components/page/secret/metadata/version-diff.js (hunk 3 above)
Comment (hellobontempo) [reply ui/lib/kv/addon/components/page/secret/metadata/version-diff.js:69]: can this action trigger the network request so instead `this.rightVersion` tracked property is the secret version data?

File: ui/lib/kv/addon/components/page/secret/metadata/version-diff.hbs (hunk 4 above)
Comment (Monkeychip) [reply ui/lib/kv/addon/components/page/secret/metadata/version-diff.hbs:45]: As discussed, a nice clean up option. Limiting how much time we're spending on this view.

File: ui/lib/kv/addon/components/page/secret/metadata/version-diff.js (hunk 3 above)
Comment (Monkeychip) [reply ui/lib/kv/addon/components/page/secret/metadata/version-diff.js:71]: Update: we explored and came to a similar pattern. Going to resolve this for ease of the next round of reviews.

File: ui/lib/kv/addon/components/page/secret/metadata/version-diff.hbs (hunk 5 above)
Comment (Monkeychip) [reply ui/lib/kv/addon/components/page/secret/metadata/version-diff.hbs]: Update: modified to match version history and version drop down. Added a is-active link for selected versions.

File: ui/lib/kv/addon/components/page/secret/metadata/version-diff.js (hunk 7 above)
Comment (Monkeychip) [reply ui/lib/kv/addon/components/page/secret/metadata/version-diff.js:66]: Update: we explored and came to a similar pattern. Going to resolve this for ease of the next round of reviews.

File: ui/app/styles/core.scss
Comment (Monkeychip) [inline ui/app/styles/core.scss:1]: This is kind of a duplicate of the old component scss file. However, when we remove the old code, the duplicate will also be removed.

File: ui/lib/kv/addon/components/page/secret/metadata/version-diff.hbs [hunk 8]
```diff
@@ -1,11 +1,107 @@
-<KvPageHeader @breadcrumbs={{@breadcrumbs}} @model={{@model}} @pageTitle={{@model.name}}>
+<KvPageHeader @breadcrumbs={{@breadcrumbs}} @model={{@model}} @pageTitle={{@path}}>
```
Comment (hellobontempo) [inline ui/lib/kv/addon/components/page/secret/metadata/version-diff.hbs:1]: I think `@model` can be removed here

**Review by hellobontempo — APPROVED** [review]
Looks good! Nice job tackling. Feel free to change the getter I refactored to something like `deactivatedState` if that makes more sense 😄

**Pull request merged**